### PR TITLE
Fixes bugs in RotationEquivariant2dCore. Extends functionality allowing to build the Ecker / Ustyuzhaninov rotation equivariant model.

### PR DESCRIPTION
**Pull request by MaxFBurg**
This PR 
* Fixes a broken computational graph in `HermiteConv2D` in `hermite.py` layers that prevented models from training
* Fixes bugs in the `RotationEquivariant2dCore` that prevented this core from training 
* Adds functionality required to build the Ecker / Ustyuzhaninov rotation equivariant model:
  * Turn off the scale in the final layer's batch norm
  * Set various nonlinearity types (in this case the added SoftThreshold activation function is required)
  * Use different number of hidden_channels in each layer (PR adds possibility to pass a list of hidden channels)
* Enables the SpikingEncoder to have its final elu nonlinearity turned off, returning (pseudo) logits (required to optimize controversial images)

All changes made are fully backwards compatible.

When working on the code, I realized that the `independent_bn_bias` was not required even before I made any changes. If you think through all possible combintations of `self.bias` and `self.batch_norm_scale` in lines 209-223 it becomes apparent that the case of `self.independent_bn_bias=True` is included in lines 211-223 (corresponding to `self.independent_bn_bias=True`), so this argument is meaningless. I added a deprecation warning to keep backwards compability and would suggest we remove it in the future.

### CODE REVIEW

**Review by mohammadbashiri — CHANGES_REQUESTED** [review]
Re `independent_bn_bias`: Do you mean that setting it to `True` or `False` has no effect on the model? Both from the code and documentation it seems that this argument would additionally define a bias (and scale) if it is set to `False`. So I am not sure if the argument has no effect on the model.

Or do you think it does not make sense to have it? To me, it is kind of weird that when `independent_bn_bias=False`, which supposedly means that bias and bn are NOT independent, there is suddenly a bias defined independently in addition to the batchnorm. But I am not familiar with the model and the code at all, so I might be missing something.

For the future fix/improvement, would you suggest removing the argument and directly using [this code block](https://github.com/MaxFBurg/neuralpredictors/blob/e29b7997be1188de3674cb9455c4df12ea974914/neuralpredictors/layers/cores/conv2d.py#L212-L223) or you had other ideas?

File: neuralpredictors/layers/activations.py [hunk 1]
```diff
@@ -186,3 +186,12 @@ def forward(self, *args, data_key=None, **kwargs):
 
     def regularizer(self, data_key):
         return self[data_key].smoothness_regularizer()
+
+
+class SoftThreshold(nn.Module):
+    def __init__(self) -> None:
+        super().__init__()
+        self.zero = torch.nn.parameter.Parameter(torch.tensor(0, dtype=torch.float32), requires_grad=False)
```
[the remark this answers is not in the document]
Comment (MaxFBurg) [reply neuralpredictors/layers/activations.py:194]: `Parameter` is in `torch.nn.parameter` (https://pytorch.org/docs/stable/generated/torch.nn.parameter.Parameter.html#torch.nn.parameter.Parameter).

Yes, always gradientless (constant). That is a great suggestion, I revised the code accordingly.

[the remark this answers is not in the document]
Comment (MaxFBurg) [reply neuralpredictors/layers/activations.py]: > Re `independent_bn_bias`: Do you mean that setting it to `True` or `False` has no effect on the model? Both from the code and documentation it seems that this argument would additionally define a bias (and scale) if it is set to `False`. So I am not sure if the argument has no effect on the model.
> 
> Or do you think it does not make sense to have it? To me, it is kind of weird that when `independent_bn_bias=False`, which supposedly means that bias and bn are NOT independent, there is suddenly a bias defined independently in addition to the batchnorm. But I am not familiar with the model and the code at all, so I might be missing something.
> 
> For the future fix/improvement, would you suggest removing the argument and directly using [this code block](https://github.com/MaxFBurg/neuralpredictors/blob/e29b7997be1188de3674cb9455c4df12ea974914/neuralpredictors/layers/cores/conv2d.py#L212-L223) or you had other ideas?

I think it does not make much sense to have this argument, so I would suggest exactly what you described in the last paragraph above. Having `independent_bn_bias` as an argument is superficial, and in some sense artificial/technical: what you are interested in is configuring batchnorm such that it can have any combination of scale=True/False and bias=True/False. In the end, it does not matter if this is done by `nn.BatchNorm2d` or by turning both off in `nn.BatchNorm2d` and then adding `Bias2DLayer` and `Scale2DLayer` (or any combination). So from the perspective of the users: they do not care how this is done internally.

If you look at the code block you cited above: this handles all combinations of bias and scale, so this code block is sufficient to configure batchnorm behavior in any way you like. (It just looks this complicated because in `nn.BatchNorm2d` you can just turn on/off both scale and bias at the same time, so if you just want to have scale xor bias, you have to turn both off and introduce a scale / bias layer yourself. In tensorflow, you could configure scale and bias in batchnorm independently). 

So bottom line: remove `independent_bn_bias`, because it is superficial / does not make sense to have it, as all cases / configurations are handled in [this code block](https://github.com/MaxFBurg/neuralpredictors/blob/e29b7997be1188de3674cb9455c4df12ea974914/neuralpredictors/layers/cores/conv2d.py#L212-L223) .

File: neuralpredictors/layers/cores/conv2d.py [hunk 2]
```diff
@@ -173,36 +201,46 @@ def set_batchnorm_type(self):
...
             else:
                 layer["norm"] = self.batchnorm_layer_cls(
-                    self.hidden_channels, momentum=self.momentum, affine=self.bias and self.batch_norm_scale
+                    hidden_channels,
+                    momentum=self.momentum,
+                    affine=self.bias
+                    and self.batch_norm_scale
+                    and (not self.pre_final_layer_built() or self.final_batchnorm_scale),
                 )
                 if self.bias:
```
Comment (MaxFBurg) [inline neuralpredictors/layers/cores/conv2d.py:219]: `and not`

**Review by KonstantinWilleke — CHANGES_REQUESTED** [review]

File: neuralpredictors/layers/cores/conv2d.py [hunk 3]
```diff
@@ -173,36 +201,47 @@ def set_batchnorm_type(self):
...
 
-    def add_bn_layer(self, layer):
+    def pre_final_layer_built(self):
+        return len(self.features) == self.num_layers - 1
+
+    def add_bn_layer(self, layer, hidden_channels):
         if self.batch_norm:
             if self.independent_bn_bias:
-                layer["norm"] = self.batchnorm_layer_cls(self.hidden_channels, momentum=self.momentum)
+                layer["norm"] = self.batchnorm_layer_cls(hidden_channels, momentum=self.momentum)
```
[the remark this answers is not in the document]
Comment (MaxFBurg) [reply neuralpredictors/layers/cores/conv2d.py:210]: Discussed it just now, is correct

File: neuralpredictors/layers/cores/conv2d.py [hunk 4]
```diff
@@ -173,36 +201,47 @@ def set_batchnorm_type(self):
...
             else:
                 layer["norm"] = self.batchnorm_layer_cls(
-                    self.hidden_channels, momentum=self.momentum, affine=self.bias and self.batch_norm_scale
+                    hidden_channels,
```
[the remark this answers is not in the document]
Comment (MaxFBurg) [reply neuralpredictors/layers/cores/conv2d.py:213]: Discussed it just now, is correct

File: neuralpredictors/layers/cores/conv2d.py [hunk 5]
```diff
@@ -173,36 +201,47 @@ def set_batchnorm_type(self):
         self.bias_layer_cls = Bias2DLayer
         self.scale_layer_cls = Scale2DLayer
 
-    def add_bn_layer(self, layer):
+    def pre_final_layer_built(self):
```
[the remark this answers is not in the document]
Comment (MaxFBurg) [reply neuralpredictors/layers/cores/conv2d.py]: That's great, thank you!

**Comment by KonstantinWilleke** [comment]
One last issue: in the `forward` of the factorized readout, there is still the issue with the `width` and `height`

https://github.com/sinzlab/neuralpredictors/blob/017bf4d01d5615f11e4d40b618122ffaed4bbb7e/neuralpredictors/layers/readouts/factorized.py#L156

**Review by KonstantinWilleke — APPROVED** [review]

**Review by mohammadbashiri — APPROVED** [review]
